language/libra-vm/src/on_chain_configs.rs

```diff
@@ -0,0 +1,67 @@
+use libra_types::{on_chain_config::access_path_for_config, transaction::SCRIPT_HASH_LENGTH};
+use move_core_types::identifier::Identifier;
+use move_vm_state::data_cache::RemoteCache;
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+
+#[derive(Debug, Clone)]
+pub(crate) struct VMConfig {
+    pub publishing_options: VMPublishingOption,
+}
+
+/// Defines and holds the publishing policies for the VM. There are three possible configurations:
+/// 1. No module publishing, only whitelisted scripts are allowed.
+/// 2. No module publishing, custom scripts are allowed.
+/// 3. Both module publishing and custom scripts are allowed.
+/// We represent these as an enum instead of a struct since whitelisting and module/script
+/// publishing are mutually exclusive options.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum VMPublishingOption {
+    /// Only allow scripts on a whitelist to be run
+    Locked(BTreeSet<[u8; SCRIPT_HASH_LENGTH]>),
+    /// Allow custom scripts, but _not_ custom module publishing
+    CustomScripts,
+    /// Allow both custom scripts and custom module publishing
+    Open,
+}
+
+impl VMConfig {
+    pub fn load_on_chain_config(state_view: &dyn RemoteCache) -> Option<Self> {
+        Some(VMConfig {
+            publishing_options: load_script_whitelist(state_view)?,
+        })
+    }
+}
+
+impl VMPublishingOption {
+    pub fn is_open(&self) -> bool {
+        match self {
+            VMPublishingOption::Open => true,
+            _ => false,
+        }
+    }
+
+    pub fn get_whitelist_set(&self) -> Option<&BTreeSet<[u8; SCRIPT_HASH_LENGTH]>> {
+        match self {
+            VMPublishingOption::Locked(whitelist) => Some(&whitelist),
+            _ => None,
+        }
+    }
+}
+
+static SCRIPT_WHITELIST: Lazy<Identifier> =
+    Lazy::new(|| Identifier::new("ScriptWhitelist").unwrap());
+
+fn load_script_whitelist(state_view: &dyn RemoteCache) -> Option<VMPublishingOption> {
```

Comment:
this probably can be made into a trait function

Reply:
Actually this is something I was thinking about doing. The main concern is that I don't know where to put this trait. `libra_types` is not ideal cuz this trait will probably needs a `StateView` to load the raw bytes from storage, and I'm not sure where else is suitable for such trait. Maybe we should have a top level directory called `libra_onchain_config` or something. What's your thought?

Reply:
I feel it's hard for other use case to have &StateView as interface since it's very execution/storage faced struct.
preliminary thought: we probably can split the interface into request/response as a storage API? sth like
```
storage API: fn fetch_config(config_path: AccessPath) -> bytes;
trait API: fn fetch_config(id: Identifier) -> T;
```